### PR TITLE
Ft implement jwt 162473931

### PR DESCRIPTION
**Pull request by Amoswachira**
#### What does this PR do?
Implements jwt (json web Token)
#### Description of Task to be completed?
Have the following endpoints working
The following is a protected route and only logged in users are allowed 
GET  /api /v2/test
#### How should this be manually tested?
Clone the repo cd into IRepoter
make virtual environment
install the requirements.txt file
Then run the flask run command
using postman navigate to the above url 
Then in the Authorization Tab select the Bearer token in the drop down 
Then in the Token field enter the access token 
Then hit send to access the protected route
#### What are the relevant pivotal tracker stories?
#162473931
#### Screenshots (if appropriate)
protected route

### CODE REVIEW

File: app/api/v2/models.py [hunk 1]
```diff
@@ -207,3 +207,42 @@ def findAll(self):
...
+    def register_user(self, username, password, email):
+        """method to sign up a user"""
+        con = self.connect()
+        cursor = con.cursor()
+        cursor.execute("SELECT username FROM users\
+                         WHERE username = %s", (username,))
+        user = cursor.fetchone()
+        if user is not None:
+            return False
+        cursor.execute("SELECT password FROM users\
```
Comment (thoth-ky) [inline app/api/v2/models.py]: don't check users passwords

File: app/api/v2/views.py [hunk 2]
```diff
@@ -53,10 +54,255 @@ def valid_characters(value):
...
+                            help="Phone number field is optional.")
+
+        args = parser.parse_args()
+        data = request.get_json(silent=True)
+        username = data["username"]
+        password = data["password"]
+        email = data["email"]
+        firstname = data["firstname"]
+        lastname = data["lastname"]
+        valid_data = db.register_user(username, password, email)
```
Comment (thoth-ky) [inline app/api/v2/views.py]:
```suggestion
        valid_data = db.validate_user_details(username, password, email)
```

File: app/api/v2/views.py [hunk 3]
```diff
@@ -53,10 +54,255 @@ def valid_characters(value):
...
+                            help="Firstname field is optional.")
+        parser.add_argument("lastname",
+                            type=str,
+                            help="Lastname field is optional.")
+        parser.add_argument("phoneNumber",
+                            type=int,
+                            help="Phone number field is optional.")
+
+        args = parser.parse_args()
+        data = request.get_json(silent=True)
```
Comment (thoth-ky) [inline app/api/v2/views.py]:
```suggestion
        data =  parser.parse_args()
```

File: app/api/v2/views.py [hunk 4]
```diff
@@ -53,10 +54,255 @@ def valid_characters(value):
...
+        username = data["username"]
+        password = data["password"]
+        email = data["email"]
+        firstname = data["firstname"]
+        lastname = data["lastname"]
+        valid_data = db.register_user(username, password, email)
+        if valid_data:
+            access_token = create_access_token(identity=password)
```
Comment (thoth-ky) [inline app/api/v2/views.py]: Don't use password to create access token, you can use email, username, uder_id. Do this is after actual registration

File: app/api/v2/views.py [hunk 5]
```diff
@@ -53,10 +54,255 @@ def valid_characters(value):
...
+            posted = (data['firstname'], data['lastname'],
+                      data['othername'], email, data['phoneNumber'],
+                      username, password)
+            db.save_user_details(posted)
+            return{"status": 201, "data":
+                   [{"token": access_token, "user": data}]}, 201
+        return {"message": "Bad credentials.Signup failed"}, 400
```
Comment (thoth-ky) [inline app/api/v2/views.py]:
```suggestion
        return {"message": "Bad credentials. Signup failed"}, 400
```

File: app/api/v2/views.py [hunk 6]
```diff
@@ -53,10 +54,255 @@ def valid_characters(value):
...
+        parser.add_argument("username",
+                            type=str,
+                            required=True,
+                            help="Username field is required.")
+        parser.add_argument("password",
+                            type=str,
+                            required=True,
+                            help="Password field is required.")
+        args = parser.parse_args()
+        data = request.get_json(silent=True)
```
Comment (thoth-ky) [inline app/api/v2/views.py]:
```suggestion
        data = parser.parse_args()
```

File: app/api/v2/views.py [hunk 7]
```diff
@@ -53,10 +54,255 @@ def valid_characters(value):
...
+        username = data["username"]
```
Comment (thoth-ky) [inline app/api/v2/views.py]:
```suggestion
        username, password = data["username"], data["password"]
```

File: app/api/v2/views.py [hunk 8]
```diff
@@ -53,10 +54,255 @@ def valid_characters(value):
...
+        password = data["password"]
+        valid_data = db.login_user(username, password)
```
Comment (thoth-ky) [inline app/api/v2/views.py]:
```suggestion
        valid_login = db.login_user(username, password)
```

File: app/api/v2/views.py [hunk 9]
```diff
@@ -53,10 +54,255 @@ def valid_characters(value):
...
+        if valid_data:
+            access_token = create_access_token(identity=password)
```
Comment (thoth-ky) [inline app/api/v2/views.py]: use username, or id or email not password here

**Review by thoth-ky — CHANGES_REQUESTED** [review]
Nice work. Kindly check suggested changes

File: app/api/v2/models.py (hunk 1 above)
Comment (Amoswachira) [reply app/api/v2/models.py]: fixed

**Pull request merged**